### PR TITLE
reuse the digest auth state to avoid unnecessary requests

### PR DESCRIPTION
**Pull request by rettier**
this implements digest authentication state reuse as discussed in #1467.

Contrary to the [comment](https://github.com/encode/httpx/issues/1467#issuecomment-1280656164) i am not skipping most of the authentication flow if a previous challenge exists. Otherwise a failed authentication (401) would not result in a retry with the new server challenge (`test_digest_auth_resets_nonce_count_after_401` would fail).

### CODE REVIEW

File: tests/client/test_auth.py [hunk 1]
```diff
@@ -527,6 +527,47 @@ async def test_digest_auth_incorrect_credentials() -> None:
...
+
+    async with httpx.AsyncClient(transport=httpx.MockTransport(app)) as client:
+        response_1 = await client.get(url, auth=auth)
+        response_2 = await client.get(url, auth=auth)
+
+        assert response_1.status_code == 200
+        assert response_2.status_code == 200
+
+        assert len(response_1.history) == 1
+        assert len(response_2.history) == 0
```
Comment (tomchristie) [inline tests/client/test_auth.py:544]: Like this test, nice and clear. 👍

File: tests/client/test_auth.py [hunk 2]
```diff
@@ -527,6 +527,47 @@ async def test_digest_auth_incorrect_credentials() -> None:
     assert len(response.history) == 1
 
 
+@pytest.mark.asyncio
+async def test_digest_auth_reuses_challenge() -> None:
+    url = "https://example.org/"
+    auth = DigestAuth(username="tomchristie", password="password123")
```
Comment (tomchristie) [inline tests/client/test_auth.py:533]: Eeek... you've included my password in clear text. 😬
I'd suggest anonymising it like this instead...

```suggestion
    auth = DigestAuth(username="user", password="password123")
```

File: tests/client/test_auth.py [hunk 3]
```diff
@@ -527,6 +527,47 @@ async def test_digest_auth_incorrect_credentials() -> None:
...
+        assert response_2.status_code == 200
+
+        assert len(response_1.history) == 1
+        assert len(response_2.history) == 0
+
+
+@pytest.mark.asyncio
+async def test_digest_auth_resets_nonce_count_after_401() -> None:
+    url = "https://example.org/"
+    auth = DigestAuth(username="tomchristie", password="password123")
```
Comment (tomchristie) [inline tests/client/test_auth.py:550]:
```suggestion
    auth = DigestAuth(username="user", password="password123")
```

File: tests/client/test_auth.py [hunk 4]
```diff
@@ -527,6 +527,47 @@ async def test_digest_auth_incorrect_credentials() -> None:
...
+    app = DigestApp()
+
+    async with httpx.AsyncClient(transport=httpx.MockTransport(app)) as client:
+        response_1 = await client.get(url, auth=auth)
+        assert response_1.status_code == 200
+        assert len(response_1.history) == 1
+        first_nonce = response_1.json()["auth"].split("nonce=")[1].split(",")[0]
+        first_nc = response_1.json()["auth"].split("nc=")[1].split(",")[0]
+
+        app.send_response_after_attempt = 2
```
Comment (tomchristie) [inline tests/client/test_auth.py:560]: This is a bit obscure to me, I think this line deserves a comment.

File: tests/test_auth.py [hunk 5]
```diff
@@ -61,3 +61,39 @@ def test_digest_auth_with_401():
     response = httpx.Response(content=b"Hello, world!", status_code=200)
     with pytest.raises(StopIteration):
         flow.send(response)
+
+
+def test_digest_auth_with_401_nonce_counting():
```
Comment (tomchristie) [inline tests/test_auth.py:66]: Very nice, yup.

**Review by tomchristie — COMMENTED** [review]
Great stuff - thanks!
A few minor suggestions here.

File: tests/client/test_auth.py (hunk 4 above)
Comment (rettier) [reply tests/client/test_auth.py:567]: added some explanation for the assertions / send_response_after

**Comment by rettier** [comment]
I have refactored some code in view of your suggestions. 
Since you removed your username from the test-data i have also removed it from the existing basic auth test cases.

Regarding the failing test pipeline i assume this is not an issue with my branch, as other pipelines seem to fail with the same error message?

**Comment by tomchristie** [comment]
> Regarding the failing test pipeline i assume this is not an issue with my branch, as other pipelines seem to fail with the same error message?

Yup. See https://github.com/encode/httpx/pull/2465

**Review by tomchristie — APPROVED** [review]

**Comment by tomchristie** [comment]
Good work, thanks!

**Pull request merged**